lnwallet/chanfunding/coin_select.go

```diff
@@ -68,53 +125,47 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
 			return nil, 0, err
 		}
 
-		var weightEstimate input.TxWeightEstimator
-
-		for _, utxo := range selectedUtxos {
-			switch {
-
-			case txscript.IsPayToWitnessPubKeyHash(utxo.PkScript):
-				weightEstimate.AddP2WKHInput()
-
-			case txscript.IsPayToScriptHash(utxo.PkScript):
-				weightEstimate.AddNestedP2WKHInput()
-
-			default:
-				return nil, 0, fmt.Errorf("unsupported address type: %x",
-					utxo.PkScript)
-			}
+		// Obtain fee estimates both with and without using a change
+		// output.
+		requiredFeeNoChange, requiredFeeWithChange, err := calculateFees(
+			selectedUtxos, feeRate)
+		if err != nil {
+			return nil, 0, err
 		}
 
-		// Channel funding multisig output is P2WSH.
-		weightEstimate.AddP2WSHOutput()
-
-		// Assume that change output is a P2WKH output.
-		//
-		// TODO: Handle wallets that generate non-witness change
-		// addresses.
-		// TODO(halseth): make coinSelect not estimate change output
-		// for dust change.
-		weightEstimate.AddP2WKHOutput()
-
 		// The difference between the selected amount and the amount
-		// requested will be used to pay fees, and generate a change
-		// output with the remaining.
+		// requested will be used to pay fees and generate a change
+		// output with the remaining if above dust limit.
 		overShootAmt := totalSat - amt
+		changeAmt := btcutil.Amount(0)
+
+		switch {
 
 		// Based on the estimated size and fee rate, if the excess
 		// amount isn't enough to pay fees, then increase the requested
 		// coin amount by the estimate required fee, performing another
 		// round of coin selection.
-		totalWeight := int64(weightEstimate.Weight())
-		requiredFee := feeRate.FeeForWeight(totalWeight)
-		if overShootAmt < requiredFee {
-			amtNeeded = amt + requiredFee
+		case overShootAmt < requiredFeeNoChange:
+			amtNeeded = amt + requiredFeeNoChange
 			continue
+
+		// If sufficient funds were selected to cover for a change
+		// output above the dust limit, we'll add the change output.
+		// Otherwise we'll go with the no change transaction.
+		case overShootAmt > requiredFeeWithChange:
+			newChange := overShootAmt - requiredFeeWithChange
+			if newChange > dustLimit {
+				changeAmt = newChange
+			}
```

Comment:
I would add a default clause to handle the case `requiredFeeNoChange < overShootAmt < requiredFeeWithChange` for readability.

```go
// If sufficient funds were selected to cover the fee required to include a change
// output, the remainder will be our change amount.
case overShootAmt > requiredFeeWithChange:
    changeAmt = overShootAmt - requiredFeeWithChange

// Otherwise we have selected enough to pay for a tx without a change output.
default:
      changeAmt = 0
}


if changeAmt < dustLimit {
    changeAmt = 0
}

Reply:
Great!